calico-vpp-agent/watchers/secret_watcher.go

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) 2019-2020 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright (C) 2022 Cisco Systems Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watchers
+
+import (
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+type secretWatchData struct {
+	// The channel that we should write to when we no longer want this watch
+	stopCh chan struct{}
+
+	// Secret value
+	secret *v1.Secret
+}
+
+type SecretWatcherClient interface {
+	// this function is invoked upon add|update|delete of a secret
+	OnSecretUpdate(secret string)
```

Comment:
Why not `OnSecretUpdate(old, new *v1.Secret)` ? If we pass both previous / new values and the whole secret object, that'll allow us to directly use the name & secret value, and behave differently depending on whether it's an addition, update or deletion

Reply:
Yes, it would make sense if we want to do very elaborate processing with them. But our usage is rather simple really, namely, re-check/re-evaluate the BGP peers when there's some update related to the secret. For all the cases, ie, addition/deletion/update, there's really no difference in the processing required imo (at the moment, at least). Its the same in calico case too afaik. We could re-visit if need be, of course.

Moreover, the pair of (old, new) objects are available only for the update case, not for addition and deletion.